gollum/gollum
# Allow for overriding only specific Mustache templates/partials.

Relates to #1450.

* This PR adds the ability to override Mustache template and partial files without having to copy the entire `lib/gollum/templates/` directory. 
* This PR introduces a change in behavior for the `--template-dir` configuration directive in that the default `Precious::App.settings.mustache[:templates]` path now **never** changes, and is instead **supplemented** with an additional directory that contains override files.

## Screenshot

<img width="1043" alt="gollum-template-dir-cascade" src="https://user-images.githubusercontent.com/637270/118331210-e9476e00-b4cd-11eb-8b0e-340dccc71659.png">

## Testing locally

* Clone and check out this branch.
* Run `bin/gollum $YOUR_LOCAL_WIKI_PATH --template-dir=./test/examples/template_cascade`
* Open http://localhost:4567/Home and note the big red sections on the page.

## Issues

* [x] Currently the single provided test passes when run in isolation (`ruby -I"lib:lib:test:." test/test_template_cascade.rb`)
  * ~~However, the test fails when run as a part of the entire suite (`bundle exec rake test`). I suspect this is due to the `Page` View maintaining state between individual test file execution but haven't figured out how to reset it explicitly.~~
  * ~~A debugging line is currently included (that should be removed before merge) to expose the template selection process.~~
* [x] There likely needs to be more tests added once the above testing issue can be solved.
